File: ios/Sources/KlinikDocumentsFeature/DocumentsModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum DocumentsPhase: Sendable, Equatable {
    case loading
    case loaded
    /// Nothing uploaded yet. Not a failure, and must not be shown as one.
    case empty
    /// Absent, or outside this user's scope — the backend makes the two
    /// indistinguishable on purpose.
    case notFound
    case failed(String)
}

public struct DocumentsState: Sendable, Equatable {
    public var phase: DocumentsPhase = .loading
    public var documents: [ClinicalDocument] = []
    public var nextCursor: String?
    public var isLoadingMore = false
    public var uploading = false
    public var uploadError: String?
    /// How much of the current upload has reached the server, for a progress bar.
    public var uploadProgress: UploadProgress?

    public var hasMore: Bool { nextCursor != nil }

    /// Whether anything is still being processed, so the screen knows whether
    /// there is any point polling again.
    public var hasUnsettledWork: Bool {
        documents.contains { !$0.ocrStatus.isSettled }
    }

    public init() {}
}

/// A patient's documents: the list, uploading, and watching processing finish.
public actor DocumentsModel {
    private let api: DocumentsAPI
    private let resumable: ResumableUpload
    private let patientId: String
    private let pageSize: Int

    /**
     * Above this, the upload is chunked and resumable.
     *
     * Below it, a failed attempt costs one small request and the three-call
     * dance would be pure overhead. Above it, a dropped connection costs the
     * patient their whole transfer — which on mobile data abroad is the case
     * this product has to survive.
     */
    private let resumableThreshold: Int

    private(set) public var state = DocumentsState()

    public init(
        api: DocumentsAPI,
        resumable: ResumableUpload,
        patientId: String,
        pageSize: Int = 25,
        resumableThreshold: Int = ResumableUpload.chunkSize
    ) {
        self.api = api
        self.resumable = resumable
        self.patientId = patientId
        self.pageSize = pageSize
        self.resumableThreshold = resumableThreshold
    }

    public func currentState() -> DocumentsState { state }

    public func load() async {
        state.phase = .loading
        state.documents = []
        state.nextCursor = nil
        await fetchPage(cursor: nil)
    }

    /// The next page. Ignored while one is already in flight, so a fast scroll
    /// does not request the same cursor twice and duplicate rows.
    public func loadMore() async {
        guard let cursor = state.nextCursor, !state.isLoadingMore else { return }

        state.isLoadingMore = true
        defer { state.isLoadingMore = false }

        await fetchPage(cursor: cursor)
    }

    /// Uploads a file and reloads the list.
    ///
    /// Reloading rather than prepending what we sent: the server decides the
    /// document's real type from the bytes, and a row showing what the client
    /// guessed would disagree with the one beside it after the next refresh.
    @discardableResult
    public func upload(
        fileURL: URL,
        type: DocumentType,
        contentType: String = "application/octet-stream"
    ) async -> Bool {
        guard !state.uploading else { return false }

        state.uploading = true
        state.uploadError = nil
        state.uploadProgress = nil
        defer {
            state.uploading = false
            state.uploadProgress = nil
        }

        do {
            if (try? fileSize(of: fileURL)) ?? 0 > resumableThreshold {
                try await sendResumable(fileURL: fileURL, type: type)
            } else {
                _ = try await api.upload(
                    patientId: patientId,
                    fileURL: fileURL,
                    type: type,
                    contentType: contentType
                )
            }
        } catch let error as APIError {
            // A refused upload is the type or size check doing its job, and the
            // server's message says which. Our own would say less.
            state.uploadError = L10n.message(for: error)
            return false
        } catch {
            state.uploadError = L10n.string("error.server")
            return false
        }

        await load()
        return true
    }

    /**
     * Re-reads the first page to pick up processing that has finished.
     *
     * Polling, not pushing: live updates arrive with the message socket in a
     * later task, and until then a screen that never updates would leave a
     * document sitting at "queued" long after it was done.
     */
    public func refreshStatuses() async {
        guard state.hasUnsettledWork else { return }

        do {
            let page = try await api.list(patientId: patientId, limit: pageSize)
            merge(page.items)
        } catch {
            // A failed poll is not worth interrupting the screen for: the list
            // on display is still correct, only slightly stale.
        }
    }

    public func remove(documentId: String) async {
        do {
            try await api.remove(documentId: documentId)
            state.documents.removeAll { $0.id == documentId }
            if state.documents.isEmpty { state.phase = .empty }
        } catch let error as APIError {
            state.uploadError = L10n.message(for: error)
        } catch {
            state.uploadError = L10n.string("error.server")
        }
    }

    /// Opens a session, streams the file, and completes it. The session id is
    /// kept for the duration so an interrupted send resumes rather than
    /// restarts; surviving an app relaunch needs the local store that T2.6
    /// still owes.
    private func sendResumable(fileURL: URL, type: DocumentType) async throws {
        let session = try await resumable.begin(
            patientId: patientId,
            type: type,
            originalName: fileURL.lastPathComponent
        )

        do {
            _ = try await resumable.send(fileURL: fileURL, sessionId: session.id) { progress in
                Task { await self.report(progress) }
            }

            _ = try await resumable.complete(sessionId: session.id, fileURL: fileURL)
        } catch {
            // Releases the parts already sent. Best-effort: the sweep on the
            // server catches whatever this misses.
            try? await resumable.abort(sessionId: session.id)
            throw error
        }
    }

    private func report(_ progress: UploadProgress) {
        state.uploadProgress = progress
    }

    private func fileSize(of fileURL: URL) throws -> Int {
        let attributes = try FileManager.default.attributesOfItem(atPath: fileURL.path)
        return (attributes[.size] as? Int) ?? 0
    }

    private func fetchPage(cursor: String?) async {
        do {
            let page = try await api.list(
                patientId: patientId,
                cursor: cursor,
                limit: pageSize
            )

            state.documents += page.items
            state.nextCursor = page.nextCursor
            state.phase = state.documents.isEmpty ? .empty : .loaded
        } catch let error as APIError {
            if case .notFound = error {
                state.phase = .notFound
            } else {
                state.phase = .failed(L10n.message(for: error))
            }
        } catch {
            state.phase = .failed(L10n.string("error.server"))
        }
    }

    /// Updates statuses in place rather than replacing the list, so a poll
    /// landing mid-scroll does not throw away pages already loaded.
    private func merge(_ fresh: [ClinicalDocument]) {
        var byId: [String: ClinicalDocument] = [:]
        for document in fresh { byId[document.id] = document }

        state.documents = state.documents.map { byId[$0.id] ?? $0 }
    }
}

File: android/feature/documents/src/main/kotlin/xyz/klinik/feature/documents/DocumentsModel.kt
package xyz.klinik.feature.documents

import java.io.File
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import xyz.klinik.network.ApiError
import xyz.klinik.network.ClinicalDocument
import xyz.klinik.network.DocumentType
import xyz.klinik.network.DocumentsApi
import xyz.klinik.network.ResumableUpload
import xyz.klinik.network.UploadProgress
import xyz.klinik.network.UiText
import xyz.klinik.network.messageKey
import xyz.klinik.network.uiText

sealed interface DocumentsPhase {
    data object Loading : DocumentsPhase
    data object Loaded : DocumentsPhase

    /** Nothing uploaded yet. Not a failure, and must not be shown as one. */
    data object Empty : DocumentsPhase

    /**
     * Absent, or outside this user's scope — the backend makes the two
     * indistinguishable on purpose.
     */
    data object NotFound : DocumentsPhase

    data class Failed(val messageKey: String) : DocumentsPhase
}

data class DocumentsState(
    val phase: DocumentsPhase = DocumentsPhase.Loading,
    val documents: List<ClinicalDocument> = emptyList(),
    val nextCursor: String? = null,
    val loadingMore: Boolean = false,
    val uploading: Boolean = false,
    /** How much of the current upload has reached the server, for a progress bar. */
    val uploadProgress: UploadProgress? = null,
    val uploadError: UiText? = null,
) {
    val hasMore: Boolean get() = nextCursor != null

    /**
     * Whether anything is still being processed, so the screen knows whether
     * there is any point asking again.
     */
    val hasUnsettledWork: Boolean get() = documents.any { !it.ocrStatus.isSettled }
}

/** A patient's documents: the list, uploading, and watching processing finish. */
class DocumentsModel(
    private val api: DocumentsApi,
    private val resumable: ResumableUpload,
    private val patientId: String,
    private val pageSize: Int = 25,
    /**
     * Above this, the upload is chunked and resumable.
     *
     * Below it, a failed attempt costs one small request and the three-call
     * dance would be pure overhead. Above it, a dropped connection costs the
     * patient their whole transfer — which on mobile data abroad is the case
     * this product has to survive.
     */
    private val resumableThreshold: Long = ResumableUpload.CHUNK_SIZE.toLong(),
) {
    private val _state = MutableStateFlow(DocumentsState())
    val state: StateFlow<DocumentsState> = _state.asStateFlow()

    /** Serialises uploads so a double tap cannot send the same scan twice. */
    private val uploadLock = Mutex()

    suspend fun load() {
        _state.value = _state.value.copy(
            phase = DocumentsPhase.Loading,
            documents = emptyList(),
            nextCursor = null,
        )
        fetchPage(null)
    }

    /**
     * The next page. Ignored while one is already in flight, so a fast scroll
     * does not request the same cursor twice and duplicate rows.
     */
    suspend fun loadMore() {
        val cursor = _state.value.nextCursor ?: return
        if (_state.value.loadingMore) return

        _state.value = _state.value.copy(loadingMore = true)
        fetchPage(cursor)
        _state.value = _state.value.copy(loadingMore = false)
    }

    /**
     * Uploads a file and reloads the list.
     *
     * Reloading rather than prepending what we sent: the server decides the
     * document's real type from the bytes, and a row showing what the client
     * guessed would disagree with the one beside it after the next refresh.
     */
    suspend fun upload(
        path: String,
        filename: String,
        type: DocumentType,
        contentType: String = "application/octet-stream",
    ): Boolean {
        if (!uploadLock.tryLock()) return false

        _state.value = _state.value.copy(
            uploading = true,
            uploadError = null,
            uploadProgress = null,
        )

        try {
            val file = File(path)

            if (file.length() > resumableThreshold) {
                sendResumable(file, filename, type)
            } else {
                api.upload(patientId, path, filename, type, contentType)
            }
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                uploading = false,
                uploadProgress = null,
                // A refused upload is the type or size check doing its job, and
                // the server's message says which.
                uploadError = (error as? ApiError)?.uiText() ?: UiText.Key("error.server"),
            )
            return false
        } finally {
            uploadLock.unlock()
        }

        load()
        _state.value = _state.value.copy(uploading = false, uploadProgress = null)
        return true
    }

    /**
     * Opens a session, streams the file, and completes it. The session id is
     * kept for the duration so an interrupted send resumes rather than
     * restarts; surviving an app relaunch needs the local store that T2.6
     * still owes.
     */
    private suspend fun sendResumable(file: File, filename: String, type: DocumentType) {
        val session = resumable.begin(patientId, type, filename)

        try {
            resumable.send(file, session.id) { progress ->
                _state.value = _state.value.copy(uploadProgress = progress)
            }

            resumable.complete(session.id, file)
        } catch (error: Throwable) {
            // Releases the parts already sent. Best-effort: the sweep on the
            // server catches whatever this misses.
            runCatching { resumable.abort(session.id) }
            throw error
        }
    }

    /**
     * Re-reads the first page to pick up processing that has finished.
     *
     * Polling, not pushing: live updates arrive with the message socket in a
     * later task, and until then a screen that never updates would leave a
     * document sitting at "waiting" long after it was done.
     */
    suspend fun refreshStatuses() {
        if (!_state.value.hasUnsettledWork) return

        val fresh = try {
            api.list(patientId, limit = pageSize).items
        } catch (error: Throwable) {
            // A failed poll is not worth interrupting the screen for: the list
            // on display is still correct, only slightly stale.
            return
        }

        val byId = fresh.associateBy { it.id }

        // Updated in place rather than replaced, so a poll landing mid-scroll
        // does not throw away pages already loaded.
        _state.value = _state.value.copy(
            documents = _state.value.documents.map { byId[it.id] ?: it },
        )
    }

    suspend fun remove(documentId: String) {
        try {
            api.remove(documentId)
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                uploadError = (error as? ApiError)?.uiText() ?: UiText.Key("error.server"),
            )
            return
        }

        val remaining = _state.value.documents.filterNot { it.id == documentId }

        _state.value = _state.value.copy(
            documents = remaining,
            phase = if (remaining.isEmpty()) DocumentsPhase.Empty else _state.value.phase,
        )
    }

    private suspend fun fetchPage(cursor: String?) {
        try {
            val page = api.list(patientId, cursor = cursor, limit = pageSize)
            val documents = _state.value.documents + page.items

            _state.value = _state.value.copy(
                documents = documents,
                nextCursor = page.nextCursor,
                phase = if (documents.isEmpty()) DocumentsPhase.Empty else DocumentsPhase.Loaded,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                phase = if (error is ApiError.NotFound) {
                    DocumentsPhase.NotFound
                } else {
                    DocumentsPhase.Failed((error as? ApiError)?.messageKey() ?: "error.server")
                },
            )
        }
    }
}
